openzl-crypto/src/poseidon/hash.rs

```diff
@@ -0,0 +1,190 @@
+    #[inline]
+    pub fn new(permutation: Permutation<S, COM>, domain_tag: S::Field) -> Self {
+        assert_eq!(ARITY + 1, S::WIDTH);
+        Self::new_unchecked(permutation, domain_tag)
+    }
+
+    /// Builds a new [`Hasher`] over `permutation` using `T` to generate the domain tag.
+    #[inline]
+    pub fn from_permutation(permutation: Permutation<S, COM>) -> Self {
+        Self::new(permutation, S::from_parameter(T::domain_tag()))
+    }
+
+    /// Computes the hash over `input` in the given `compiler` and returns the untruncated state.
+    #[inline]
+    pub fn hash_untruncated(&self, input: [&S::Field; ARITY], compiler: &mut COM) -> Vec<S::Field> {
+        let mut state = self.permutation.first_round_with_domain_tag_unchecked(
+            &self.domain_tag,
+            input,
+            compiler,
+        );
+        self.permutation
+            .permute_without_first_round(&mut state, compiler);
+        state.0.into_vec()
+    }
+}
+
+impl<S, T, const ARITY: usize, COM> Constant<COM> for Hasher<S, T, ARITY, COM>
+where
+    S: Specification<COM> + Constant<COM>,
+    S::Type: Specification<ParameterField = Const<S::ParameterField, COM>>,
+    S::ParameterField: Constant<COM>,
+    T: DomainTag<S> + Constant<COM>,
+    T::Type: DomainTag<S::Type>,
+{
+    type Type = Hasher<S::Type, T::Type, ARITY>;
+
+    #[inline]
+    fn new_constant(this: &Self::Type, compiler: &mut COM) -> Self {
+        Self::from_permutation(this.permutation.as_constant(compiler))
+    }
+}
+
+impl<S, T, const ARITY: usize, COM> ArrayHashFunction<ARITY, COM> for Hasher<S, T, ARITY, COM>
+where
+    S: Specification<COM>,
+    T: DomainTag<S>,
+{
+    type Input = S::Field;
+    type Output = S::Field;
+
+    #[inline]
+    fn hash(&self, input: [&Self::Input; ARITY], compiler: &mut COM) -> Self::Output {
+        self.hash_untruncated(input, compiler).take_first()
+    }
+}
+
+impl<S, T, const ARITY: usize, COM> Decode for Hasher<S, T, ARITY, COM>
+where
+    S: Specification<COM>,
+    S::Field: Decode,
+    S::ParameterField: Decode<Error = <S::Field as Decode>::Error>,
+    T: DomainTag<S>,
+{
+    type Error = <S::Field as Decode>::Error;
+
+    #[inline]
+    fn decode<R>(mut reader: R) -> Result<Self, DecodeError<R::Error, Self::Error>>
+    where
+        R: Read,
+    {
+        Ok(Self::new(
+            Decode::decode(&mut reader)?,
+            Decode::decode(&mut reader)?,
+        ))
+    }
+}
+
+impl<S, T, const ARITY: usize, COM> Encode for Hasher<S, T, ARITY, COM>
+where
+    S: Specification<COM>,
+    S::Field: Encode,
+    S::ParameterField: Encode,
+    T: DomainTag<S>,
+{
+    #[inline]
+    fn encode<W>(&self, mut writer: W) -> Result<(), W::Error>
+    where
+        W: Write,
+    {
+        self.permutation.encode(&mut writer)?;
+        self.domain_tag.encode(&mut writer)?;
+        Ok(())
+    }
+}
+
+impl<S, T, const ARITY: usize, COM> Sample for Hasher<S, T, ARITY, COM>
```

Comment:
Do you mean in order to be able to randomly sample the domain tag? Because keep in mind that generating the permutation for this `Hasher` is deterministic, so there's no need to sample that part from a generic `D`.

Reply:
It's unclear to me why we're using `Sample` though, isn't this more like `Default`?